### PR TITLE
chore: handle tag prefix

### PR DESCRIPTION
**Pull request by anis-campos**
- feat(git): handle multi service repo
- test(git): add coverage

### CODE REVIEW

File: changelog_generator/repository_manager.py [hunk 1]
```diff
@@ -9,6 +10,10 @@
     r"^v?(?P<major>\d+)[-\.]+(?P<minor>\d+)[-\.]+((?P<bug>\d+)|rc(?P<rc>\d+)){0,1}$"
 )
 
+service_tag_version_re = re.compile(
+    r"^(?P<service>\w+)/(?P<major>\d+)[-\.]+(?P<minor>\d+)[-\.]+((?P<bug>\d+)|rc(?P<rc>\d+)){0,1}$"
```
Comment (sebastien-boulle) [inline changelog_generator/repository_manager.py]: s/service/prefix

File: changelog_generator/repository_manager.py [hunk 2]
```diff
@@ -17,19 +22,42 @@
...
     res = tag_version_re.match(tag)
-    if not res:
-        return False
-    return not res.group("rc")
+    return res and not res.group("rc")
 
 
-def get_tag_value(tag: str) -> int:
+def get_tag_value(tag: str) -> Tuple:
+    """This splits a tag in a tuple of ints. This makes it straightforward to compare two tags."""
```
Comment (sebastien-boulle) [inline changelog_generator/repository_manager.py]: Indeed comparing tuple is easier, someday I should learn python :)
Comment is useless and function should be renamed to `get_semver_from_tag`

File: changelog_generator/repository_manager.py (hunk 1 above)
Comment (sebastien-boulle) [reply changelog_generator/repository_manager.py]: + previous regex one become useless with
```
     r"^(?:(?P<service>\w+)/)?(?P<major>\d+)[-\.]+(?P<minor>\d+)[-\.]+((?P<bug>\d+)|rc(?P<rc>\d+)){0,1}$"
```

File: changelog_generator/repository_manager.py [hunk 3]
```diff
@@ -17,19 +22,42 @@
...
+                res.group("major"),
+                res.group("minor"),
+                res.group("bug") or 0,
+                res.group("rc") or 0,
+            ],
+        )
+    )
+
+
+def get_service_tag_value(tag: str) -> Tuple:
```
Comment (sebastien-boulle) [inline changelog_generator/repository_manager.py]: useless function

File: changelog_generator/repository_manager.py [hunk 4]
```diff
@@ -38,64 +66,62 @@ class RepositoryManager:
     organization: str = ""
     name: str = ""
 
-    def __init__(self, uri: str) -> None:
+    def __init__(self, uri: str, service: str = "") -> None:
```
Comment (sebastien-boulle) [inline changelog_generator/repository_manager.py]: s/service/prefix

File: changelog_generator/repository_manager.py [hunk 5]
```diff
@@ -38,64 +66,62 @@ class RepositoryManager:
...
+        self.is_multi_service = bool(service)
```
Comment (sebastien-boulle) [inline changelog_generator/repository_manager.py]: is_tag_prefixed

File: changelog_generator/repository_manager.py [hunk 6]
```diff
@@ -38,64 +66,62 @@ class RepositoryManager:
...
             self.name = res.group("repository")
 
-        self.tags = self.get_tags()
-
-    def get_tags(self) -> List[str]:
-        if not self.tag_names:
-            tags: List[str] = self.repository.git.tag("--merged").split("\n")
-            self.tag_names = list(
-                sorted(filter(is_tag_version, tags), key=get_tag_value)
+    def is_service_tag_version(self, tag: str) -> bool:
```
Comment (sebastien-boulle) [inline changelog_generator/repository_manager.py]: useless

File: changelog_generator/repository_manager.py [hunk 7]
```diff
@@ -38,64 +66,62 @@ class RepositoryManager:
     organization: str = ""
     name: str = ""
 
-    def __init__(self, uri: str) -> None:
+    def __init__(self, uri: str, service: str = "") -> None:
+        self.is_multi_service = bool(service)
+        self.service = service
```
Comment (sebastien-boulle) [inline changelog_generator/repository_manager.py]: prefix

File: changelog_generator/repository_manager.py [hunk 8]
```diff
@@ -38,64 +66,62 @@ class RepositoryManager:
...
-            self.tag_names = list(
-                sorted(filter(is_tag_version, tags), key=get_tag_value)
+    def is_service_tag_version(self, tag: str) -> bool:
+        res = service_tag_version_re.match(tag)
+        return res and res.group("service") == self.service and not res.group("rc")
+
+    @property
+    @lru_cache()
+    def tags(self) -> List[str]:
+        tags: List[str] = self.repository.git.tag("--merged").split("\n")
```
Comment (sebastien-boulle) [inline changelog_generator/repository_manager.py]: This should od the trick:
```
  tags = self.repository.git.tag(f"--merged '{self.prefix}*'").split("\n")
```

**Review by sebastien-boulle — CHANGES_REQUESTED** [review]
Some minor changes. Only issue is there is no service notion here, only a tag prefix.

File: changelog_generator/__main__.py [hunk 9]
```diff
@@ -53,10 +55,10 @@ def get_commit_but_types(
 
 
 def main() -> None:
-    repository = RepositoryManager("./")
+    repository = RepositoryManager("./", os.environ.get("SERVICE"))
```
Comment (sebastien-boulle) [inline changelog_generator/__main__.py]: BTW this cannot be a env variable, it should be an arguments give the binary.

Comment (anis-campos) [reply changelog_generator/__main__.py]: I'm pretty sure that that github action allows environment variables to be set. something like 
```yaml
....
    env:
      - MY_VAR: my value
```

Comment (sebastien-boulle) [reply changelog_generator/__main__.py]: Indeed, we can use it this way.

Comment (sebastien-boulle) [reply changelog_generator/__main__.py]: BTW you have to update the readme :)

File: changelog_generator/repository_manager.py [hunk 10]
```diff
@@ -1,101 +1,156 @@
...
+
+class PrefixedTagManager(BaseTagManager):
+    """
+    Matches tags of the kind:
+        - toto/1.2.3
+        - toto/1.2.3rc1
+    """
+
+    PATTERN = re.compile(
+        r"^(?P<prefix>\w+)/(?P<major>\d+)\.(?P<minor>\d+)\.((?P<bug>\d+)|rc(?P<rc>\d+))?$"
```
Comment (sebastien-boulle) [inline changelog_generator/repository_manager.py]: Why not simply using an optional group ?
```
r"^(?:(?P<prefix>\w+)/)?(?P<major>\d+)\.(?P<minor>\d+)\.((?P<bug>\d+)|rc(?P<rc>\d+))?$"
```

Comment (anis-campos) [reply changelog_generator/repository_manager.py]: I find it easier to maintain to have classes. If someone wants to add a new kind of tag parsing/filtering/sorting, he will just have to add a other class, and the rest should still work

**Review by sebastien-boulle — APPROVED** [review]

**Pull request merged**